src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -463,36 +483,45 @@ protected Task<Solution> PostProcessChangesAsync(Solution changedSolution, Cance
+            .ToImmutableArray();
+
+        var globalOptions = changedSolution.Services.GetService<ILegacyGlobalCleanCodeGenerationOptionsWorkspaceService>();
+        var fallbackOptions = globalOptions?.Provider ?? CodeActionOptions.DefaultProvider;
+        var documentIdsAndOptions = await GetDocumentIdsAndOptionsAsync(
+            changedSolution, fallbackOptions, documentIds, cancellationToken).ConfigureAwait(false);
+
+        // Do an initial pass where we cleanup syntax.
+        var syntaxCleanedSolution = await ProducerConsumer<Document>.RunParallelAsync(
+            source: documentIdsAndOptions,
+            produceItems: static async (documentIdAndOptions, callback, changedSolution, cancellationToken) =>
             {
-                var document = processedSolution.GetRequiredDocument(documentId);
-                var processedDocument = await PostProcessChangesAsync(document, cancellationToken).ConfigureAwait(false);
-                processedSolution = processedDocument.Project.Solution;
-            }
-        }
+                var document = changedSolution.GetRequiredDocument(documentIdAndOptions.documentId);
+                var newDoc = await CleanupSyntaxAsync(document, documentIdAndOptions.codeCleanupOptions, cancellationToken).ConfigureAwait(false);
+            },
+            consumeItems: static async (stream, changedSolution, cancellationToken) =>
+            {
+                var currentSolution = changedSolution;
+                await foreach (var document in stream)
+                {
+                    currentSolution = document.SupportsSyntaxTree
+                        ? currentSolution.WithDocumentSyntaxRoot(document.Id, await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false))
+                        : currentSolution.WithDocumentText(document.Id, await document.GetValueTextAsync(cancellationToken).ConfigureAwait(false));
+                }
 
-        // process completely new projects too
-        foreach (var addedProject in solutionChanges.GetAddedProjects())
-        {
-            var documentsToProcess = addedProject.DocumentIds;
+                return currentSolution;
+            },
+            args: changedSolution,
+            cancellationToken).ConfigureAwait(false);
 
-            foreach (var documentId in documentsToProcess)
-            {
-                var document = processedSolution.GetRequiredDocument(documentId);
-                var processedDocument = await PostProcessChangesAsync(document, cancellationToken).ConfigureAwait(false);
-                processedSolution = processedDocument.Project.Solution;
-            }
-        }
+        // Then do a pass where we cleanup semantics.
+        var semanticCleanedSolution = await CleanupSemanticsAsync(
+            syntaxCleanedSolution, documentIdsAndOptions, CodeAnalysisProgress.None, cancellationToken).ConfigureAwait(false);
```

Comment:
for features that produce a new solution (as oppposed to new document) they now get the benefit of this cleanup in parallel, and only 4 passes as well across all documents.